kuma/javascript/ssr-server.js

```diff
@@ -0,0 +1,90 @@
+/* eslint-disable no-console */
+/*
+ * This is a simple Express server for performing server side rendering
+ * of our React UI. The main endpoint is POST /ssr, and it invokes the
+ * ssr() function from dist/ssr.js (which is webpacked from src/ssr.jsx).
+ */
+const express = require('express'); // Express server framework
+const morgan = require('morgan');
+
+const ssr = require('./dist/ssr.js'); // Function to do server-side rendering
+
+// Configuration
+const PID = process.pid;
+const PORT = parseInt(process.env.SSR_PORT) || 8000;
+const MAX_BODY_SIZE = 1024 * 1024; // 1 megabyte max JSON payload
+
+// Start New Relic logging if it is configured
+if (process.env.NEW_RELIC_LICENSE_KEY && process.env.NEW_RELIC_APP_NAME) {
+    console.log('Starting New Relic logging for KumaScript.');
+    require('newrelic');
+}
+
+// We're using the Express server framework
+const app = express();
+
+// Log all requests, so we get timing data for SSR.
+app.use(morgan('tiny'));
+
+// Handle JSON payloads in POST requests, putting data in req.body
+app.use(express.json({ limit: MAX_BODY_SIZE }));
+
+// A hello world endpoint for easy verification that the service is running
+app.get('/', (req, res) => {
+    res.send('<html><body><p>SSR server ready</p></body></html>');
+});
+
+// revision, health and readiness endpoints used (I think) by Kubernetes
+app.get('/revision/?', (req, res) => {
+    res.set({ 'Content-Type': 'text/plain; charset=utf-8' }).send(
+        process.env.REVISION_HASH || 'undefined'
+    );
+});
+app.get('/healthz/?', (req, res) => {
+    res.sendStatus(204);
+});
+app.get('/readiness/?', (req, res) => {
+    res.sendStatus(204);
+});
+
+// This is the main endpoint that receives document api data in the
+// request body and returns rendered HTML in the response
+app.post('/ssr/?', (req, res) => {
+    res.set({ 'Content-Type': 'text/plain; charset=utf-8' }).send(
```

Comment:
It probably doesn't matter, but I'm curious why `text/plain` rather than `text/html`? Is it because it's not a full HTML document, but just a fragment?

Reply:
I was worried that if I used text/html there might be code that would try to escape it or add a doctype or a body tag or something. If you do a client side xhr request for html, I think there is an option that causes it to be automatically parsed into a DOM tree, or something, and I wanted to be sure that nothing like that would happen.

Probably not necessary, but it seemed vaguely safer...

I'll add a comment documenting this and explaining the rationale.